deepcell_tracking/utils.py

```diff
@@ -330,3 +339,312 @@ def trks_stats(filename):
     print('Total number of divisions                  - ', total_divisions)
     print('Average cell density (cells/100 sq pixels) - ', avg_cells_per_sq_pixel * 100)
     print('Average number of frames per track         - ', int(avg_num_frames_per_track))
+
+
+def get_max_cells(y):
+    """Helper function for finding the maximum number of cells in a frame of a movie, across
+    all frames of the movie. Can be used for batches/tracks interchangeably with frames/cells.
+
+    Args:
+        y (np.array): Annotated image data
+
+    Returns:
+        max_cells (int): The maximum number of cells in any frame
+    """
+    max_cells = 0
+    for frame in range(y.shape[0]):
+        cells = np.unique(y[frame])
+        n_cells = cells[cells != 0].shape[0]
+        if n_cells > max_cells:
+            max_cells = n_cells
+    return max_cells
+
+
+def normalize_adj_matrix(adj, epsilon=1e-5):
+    # Normalize the adjacency matrix
+
+    normed_adj = np.zeros(adj.shape, dtype='float32')
+    for t in range(adj.shape[-1]):
+        adj_frame = adj[..., t]
+        # setup degree matrix
+        degree_matrix = np.zeros(adj_frame.shape, dtype=np.float32)
+        # determine whether multiple batches being normalized
+        if len(adj.shape) == 4:
+            # adj is (batch, node, node, time)
+            degrees = np.sum(adj_frame, axis=1)
+            for batch, degree in enumerate(degrees):
+                degree = (degree + epsilon) ** -0.5
+                degree_matrix[batch] = np.diagflat(degree)
+
+        elif len(adj.shape) == 3:
+            # adj is (cells, cells, time)
+            norm_adj = np.matmul(degree_matrix, adj_frame)
+            norm_adj = np.matmul(norm_adj, degree_matrix)
+            normed_adj[..., t] = norm_adj
+
+        else:
+            raise ValueError('Only 3 & 4 dim adjacency matrices are supported')
+
+    return normed_adj
+
+
+# TODO: This class contains code that could be reused for tracking.py
+#       The only difference is usually doing things by batch vs frame
```

Comment:
I would prefer to not have duplicate code in `tracking.py` if possible. We should brainstorm the best way to make the code more reusable and flexible.